TypeCobol/Compiler/CompilationUnit.cs

```diff
@@ -317,18 +318,39 @@ public void RefreshCodeAnalysisDocumentSnapshot()
                     }
 
                     //Create updated snapshot
-                    CodeAnalysisDocumentSnapshot = new InspectedProgramClassDocument(programClassDocument, diagnostics, results);
+                    CodeAnalysisDocumentSnapshot = new InspectedProgramClassDocument(programClassDocument, version, diagnostics, results);
+                    documentUpdated = true;
 
                     PerfStatsForCodeQualityCheck.OnStopRefresh();
                 }
 
-                bool CodeAnalysisDocumentNeedsUpdate()
+                bool CodeAnalysisDocumentNeedsUpdate(out int newVersion)
                 {
-                    return CodeAnalysisDocumentSnapshot == null //Not yet computed
-                           ||
-                           (CodeAnalysisDocumentSnapshot.PreviousStepSnapshot.CurrentVersion != programClassDocument.CurrentVersion); //Obsolete version
+                    if (CodeAnalysisDocumentSnapshot == null)
+                    {
+                        //Not yet computed
+                        newVersion = 0;
+                        return true;
+                    }
+
+                    if (CodeAnalysisDocumentSnapshot.PreviousStepSnapshot.CurrentVersion != programClassDocument.CurrentVersion)
+                    {
+                        //Obsolete version
+                        newVersion = CodeAnalysisDocumentSnapshot.CurrentVersion + 1;
+                        return true;
+                    }
+
+                    //No need for update
+                    newVersion = -1;
+                    return false;
                 }
             }
+
+            EventHandler<ProgramClassEvent> codeAnalysisCompleted = CodeAnalysisCompleted;
+            if (documentUpdated && codeAnalysisCompleted != null)
+            {
+                codeAnalysisCompleted(this, new ProgramClassEvent() { Version = CodeAnalysisDocumentSnapshot.CurrentVersion });
```

Comment:
This is standard practice in .NET for events because of multi-threading. Event may be unsubscribed in another thread between nullity check and actual call.

Reply:
Ok, in a multi-threaded context it makes sense!